### PR TITLE
Expiration improvements and fixes

### PR DESCRIPTION
**Pull request by jonflynng**
**Changes:**

- Added logging of soft opt-in updates to the `SoftOptInsLoggingTable`
- Added unit tests for the `deleteLink` function
- Added appropriate permissions to the lambda and simplified cloudformation by removing the `DeleteUserLinkLambdasRole`.
- Optimised dynamo queries by only calling the `getUserLinks` function once per `subscriptionId`.
- Added alarm for soft opt-ins and error handling

- [x] Tested in DEV

### CODE REVIEW

File: typescript/src/soft-opt-ins/acquisitions.ts [hunk 1]
```diff
@@ -29,7 +29,7 @@ async function updateDynamoLoggingTable(subcriptionId: string, identityId: strin
     }
 }
 
-async function handleError(identityId: string, message: string): Promise<never> {
```
Comment (jonflynng) [inline typescript/src/soft-opt-ins/acquisitions.ts:32]: this parameter wasn't being used so have deleted it and have improved error messages below

File: typescript/src/utils/aws.ts [hunk 2]
```diff
@@ -126,7 +126,14 @@ export function sendToSqs(queueUrl: string, event: any, delaySeconds?: number):
         DelaySeconds: delaySeconds
     }).promise()
 }
-export async function sendToSqsSoftOptIns(queueUrl: string, event: any, delaySeconds?: number): Promise<PromiseResult<Sqs.SendMessageResult, AWSError>> {
+
+interface SoftOptInEvent {
```
Comment (jonflynng) [inline typescript/src/utils/aws.ts:130]: Added type safety

File: cloudformation.yaml [hunk 3]
```diff
@@ -1025,6 +979,52 @@ Resources:
         Stage: !Ref Stage
         Stack: !Ref Stack
         App: !Ref App
+      Policies:
```
Comment (andrewHEguardian) [inline cloudformation.yaml:982]: just curious why did you move these from a role to policies?

File: typescript/src/link/deleteLink.ts [hunk 4]
```diff
@@ -1,51 +1,69 @@
...
+
+async function updateDynamoLoggingTable(subcriptionId: string, identityId: string) {
+    const timestamp = new Date().getTime();
+    const record = new SoftOptInLog(identityId, subcriptionId, timestamp, "Soft opt-ins processed for expired subscription");
+
+    try {
+        await dynamoMapper.put({item: record});
+        console.log(`Logged soft opt-in setting to Dynamo`);
+    } catch (error) {
+        handleSoftOptInsError(`Dynamo write failed for record: ${record}`);
```
Comment (andrewHEguardian) [inline typescript/src/link/deleteLink.ts:17]: should we log the error?

Comment (jonflynng) [reply typescript/src/link/deleteLink.ts]: yep, have added these in.

File: cloudformation.yaml (hunk 3 above)
Comment (jonflynng) [reply cloudformation.yaml:982]: Simplicity in the cloudformation. Rather than having two resources, an IAM role and a lambda I think it's simpler to just have all the permissions under the lambda.

File: typescript/src/link/deleteLink.ts [hunk 5]
```diff
@@ -60,24 +78,31 @@ export async function handler(event: DynamoDBStreamEvent): Promise<any> {
...
     let records = 0;
     let rows = 0;
-    for await (const deletedCount of subscriptionsPromises) {
+
+    const featureFlag = false;
+
+    for (const subscriptionId of subscriptionIds) {
+        const userLinksIterator = await getUserLinks(subscriptionId);
+
+        const userSubscriptions: ReadUserSubscription[] = [];
```
[the remark this answers is not in the document]
Comment (andrewHEguardian) [reply typescript/src/link/deleteLink.ts:89]: I'm a bit confused, how are we only calling the `getUserLinks` function once after this change?

[the remark this answers is not in the document]
Comment (jonflynng) [reply typescript/src/link/deleteLink.ts:89]: Ah, once per `subscriptionId` that is. Before it was being called twice per `subscriptionId`.

File: typescript/tests/link/deleteLink.test.ts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+        const result = await handler(event);
+
+        expect(mockDataMapper.query).toHaveBeenCalledTimes(1);
+        expect(mockDataMapper.query).toHaveBeenCalledWith(ReadUserSubscription, { subscriptionId: "1" }, { indexName: "subscriptionId-userId" });
+
+        expect(mockDataMapper.delete).toHaveBeenCalledTimes(1);
+        expect(mockDataMapper.delete).toHaveBeenCalledWith({"subscriptionId": "1", "userId": "123"});
+
+        /* Following is commented out as the feature flag for soft opt-ins is switched off:
```
Comment (andrewHEguardian) [inline typescript/tests/link/deleteLink.test.ts:121]: Would be nice to have this independent of the feature flag in the lambda, ie to be able to set it on in the test

**Review by andrewHEguardian — APPROVED** [review]
Looks good 👍 nice refactor

**Review by shtukas — APPROVED** [review]
